Add `cads_error_summary` method

leeky opened the pull request:
This uses the design system's `ErrorSummary` component, with a simple mapping between the auto-generated HTML ids and validation error messages from ActiveModel and the href/message format that the component expects.

![Screenshot 2024-06-17 at 16 04 15](https://github.com/citizensadvice/rails-form-builder/assets/3166/d47e4017-d4f7-4a9c-a037-5c684c16d964)

This PR now includes changes to date inputs and radio button/checkbox collections to make the give the first input or these multi-input form elements an HTML `id` that matches the single-input elements. This means we can correctly link the error summaries correctly.

I've also included some minor fixes, such as rendering some missing error messages and including a hidden field alongside checkbox groups, so that the state is correctly saved when nothing is checked.

leeky committed:
Adding `cads_error_summary`

This uses the design system's `ErrorSummary` component, with a simple
mapping between the auto-generated HTML ids and validation error
messages from ActiveModel and the href/message format that the component
expects.

marianayovcheva commented on lib/citizens_advice_form_builder/elements/error_summary.rb:
```diff
@@ -0,0 +1,29 @@
+      def error_messages
+        object.errors.group_by_attribute.map do |attr, errors|
+          {
+            href: "##{field_id_for(attr)}-input",
```
When there are multiple input fields part of a fieldset - e.g radio, checkboxes, dates, the error summary should scroll to the first input field. Not sure if this is a design system problem or if it can be adjusted here

leeky replied:
The design system uses some JS code to find the nearest label or legend to focus. It's currently broken for data fields from the looks of things, and probably needs to support other inputs too, maybe.

It's tricky for the error summary code to know if the attribute with the error was driven by a text field or a radio button in order to know which HTML `id` to target.

marianayovcheva replied:
I think the JS code is working, but it uses the href to match to the id of the input field, which in our case would end in `-input` and that's different for non textarea/text input fields.

> It's tricky for the error summary code to know if the attribute with the error was driven by a text field or a radio button in order to know which HTML id to target.

Agree. Will there be a way to do that on the form side of things? We do it for template letters but the example there is quite different since we use CMS data: https://github.com/citizensadvice/public-website/blob/a5304b890c1f0be832e4fca7e364562b795111fa/app/models/template_letters/letter.rb#L58

leeky replied:
Yeah, good point. I'm pretty sure that the GOVUK form builder gets round this by making the first radio or checkbox use the same `id` as the equivalent text field, so it's consistent.

Whilst it's possible to do what template letters is doing, by having special code which changes the `id` being targeted based on the type of form element, but I'd much prefer find a way to not need to do this.

marianayovcheva commented:
Something I noticed while testing locally - when I click on the link of the error summary it scrolls me to the the input field and passes the label. We have addressed the issue in the design system: https://github.com/citizensadvice/design-system/pull/2156/files. If I add `import initErrorSummary from '@citizensadvice/design-system/lib/error-summary';` in the form, then I get the expected behaivour - i.e scroll to the label. Wanted to check if we should add it in the form app or if there is a way to add it to the gem?

leeky commented:
@marianayovcheva Yeah, you need the ErrorSummary component's JS code in order to make it work fully. I don't think the form builder is the right place to do this (it doesn't get involved at all with the asset pipeline), as it's a progressive enhancement on top of whatever HTML gets rendered.

Probably better solved in documentation, when I get round to adding some(!)

leeky committed:
Fix form controls that didn't work with error summary

In order for `cads_error_summary` to infer the HTML IDs of the elements
to target, they need to use a common naming scheme.

Now the first element in radio button and checkbox groups, as well as
the "day" field in date input, also use the same naming scheme.

Additionally, I've added the missing error handling from some of these
fields, so when a validation error occurs, the correct styling will be
applied to them.

leeky committed:
Move common methods into Collections::Base

There were quite a few methods that were duplicated across the
collection classes for selects, radio buttons and checkbox groups, so
I've introduced a new `Collections::Base` class to hold those.

leeky committed:
Move common methods into Elements::Base

Some of the methods a consolidated into Collection::Base could be used
outside of the collection classes, so they've been moved into the
Elements::Base.

leeky committed:
Checkbox collections include hidden field

Add a hidden field with an empty value to checkbox collections. This
allows the state where no checkboxes are checked to be saved correctly.

Without this, with nothing checked, the field isn't included in the form
data and is left in its previous state.

marianayovcheva reviewed (approved):
🎉
